src/rust/graph-descriptions/src/ip_address.rs

```diff
@@ -87,4 +69,61 @@ impl NodeT for IpAddress {
     fn merge_into(&mut self, other: Self) -> bool {
         self.merge(&other)
     }
+
+    fn attach_predicates_to_mutation_unit(&self, mutation_unit: &mut MutationUnit) {
+        mutation_unit.predicate_ref("node_key", MutationPredicateValue::string(&self.node_key));
+        mutation_unit.predicate_ref(
+            "ip_address",
+            MutationPredicateValue::string(&self.ip_address),
+        );
+        mutation_unit.predicate_ref("dgraph.type", MutationPredicateValue::string("IpAddress"));
+
+        if self.first_seen_timestamp != 0 {
+            mutation_unit.predicate_ref(
+                "first_seen_timestamp",
+                MutationPredicateValue::Number(self.first_seen_timestamp as i64),
+            );
+        }
+
+        if self.last_seen_timestamp != 0 {
+            mutation_unit.predicate_ref(
+                "last_seen_timestamp",
+                MutationPredicateValue::Number(self.last_seen_timestamp as i64),
+            );
+        }
+    }
+
+    fn get_cache_identities_for_predicates(&self) -> Vec<Vec<u8>> {
+        let mut predicate_cache_identities = Vec::new();
+
+        predicate_cache_identities.push(format!(
+            "{}:{}:{}",
+            self.get_node_key(),
+            "ip_address",
+            self.ip_address
+        ));
+
+        if self.first_seen_timestamp != 0 {
+            predicate_cache_identities.push(format!(
+                "{}:{}:{}",
+                self.get_node_key(),
+                "first_seen_timestamp",
+                self.first_seen_timestamp
+            ));
+        }
+
+        if self.last_seen_timestamp != 0 {
+            predicate_cache_identities.push(format!(
+                "{}:{}:{}",
+                self.get_node_key(),
+                "last_seen_timestamp",
+                self.last_seen_timestamp
+            ));
+        }
+
+        predicate_cache_identities
+            .into_iter()
+            .map(|item| item.as_bytes().to_vec())
```

Comment:
`into_bytes` is what you want, and it's free. The code, as written, copies the bytes to a vec but we can just move the bytes right out of the String - and since a String is just a Vec<u8> with some extra guarantees, yeah, free.

Reply:
oh perfect

thanks!